src/controller/deletecontent.js

```diff
@@ -65,31 +64,56 @@ export default function deleteContent( selection, batch, options = {} ) {
 	endPos.detach();
 }
 
+// This function is a result of reaching the Ballmer's peak for just the right amount of time.
+// Even I had troubles documenting it after a while and after reading it again I couldn't believe that it really works.
 function mergeBranches( batch, startPos, endPos ) {
-	const endPath = endPos.path;
-	const mergeEnd = Math.min( startPos.path.length - 1, endPath.length - 1 );
-	let mergeDepth = compareArrays( startPos.path, endPath );
+	const startParent = startPos.parent;
+	const endParent = endPos.parent;
 
-	if ( typeof mergeDepth == 'number' ) {
-		for ( ; mergeDepth < mergeEnd; mergeDepth++ ) {
-			const mergePath = startPos.path.slice( 0, mergeDepth );
-			mergePath.push( startPos.path[ mergeDepth ] + 1 );
+	// If both positions ended up in the same parent, then there's nothing more to merge:
+	// <$root><p>x[]</p><p>{}y</p></$root> => <$root><p>xy</p>[]{}</$root>
+	if ( startParent == endParent ) {
+		return;
+	}
 
-			const mergePos = new Position( endPos.root, mergePath );
-			const previousNode = mergePos.nodeBefore;
-			const nextNode = mergePos.nodeAfter;
+	// If one of the positions is a root, then there's nothing more to merge (at least in the current state of implementation).
+	// Theoretically in this case we could unwrap the <p>: <$root>x[]<p>[]y</p></$root>, but we don't need to support it yet
+	// so let's just abort.
+	if ( !startParent.parent || !endParent.parent ) {
+		return;
+	}
 
-			if ( !checkCanBeMerged( previousNode ) || !checkCanBeMerged( nextNode ) ) {
-				return;
-			}
+	// Check if operations we'll need to do won't need to cross object or limit boundaries.
+	// E.g., we can't merge endParent into startParent in this case:
+	// <limit><startParent>x</startParent></limit><endParent></endParent>
+	if ( !checkCanBeMerged( startParent, endParent ) ) {
```

Comment:
Yep, we do remove the content, but we do touch the limit/object elements. So they can't be merged nor split.

Note: situations with selection crossing limit elements should not happen. This should be blocked on the model level. However, I wanted to have this covered at least partially here. I say partially because let's imagine this:

```html
<paragraph>x[x</paragraph>
<widget>
    noneditabletext
    <caption>x]x</caption>
</widget>
```

The expected result on delete is:


```html
<paragraph>x[]</paragraph>
<widget>
    noneditabletext
    <caption>x</caption>
</widget>
```

But you'll get:


```html
<paragraph>x[]</paragraph>
<widget>
    <caption>x</caption>
</widget>
```

So, the result is still bad, but we're at least a bit closer to the correct behaviour. The remaining obstacle is that being a limit element doesn't actually say that content within you isn't editable. So, we'll need some more element types in the schema to resolve this completely. That's why I thought that with the current types it'd be best to have the non-merging behaviour.